frontend-build-pipeline.yml

```diff
@@ -16,27 +19,48 @@ parameters:
 - name: sonarCloudProjectName # sonar cloud's cli project name
   type: string
   default: 'RevatureShopFrontend'
+- name: sonarCloudExclude # files for sonar cloud to ignore
+  type: string
+  default: '**/*.spec.ts'
+- name: sonarCloudCoverageDir
+  type: string
+  default: '$(Build.Repository.LocalPath)/coverage/lcov.info'
+- name: sonarCloudSources
+  type: string
+  default: '.'
 
 steps:
 
-
-
 # Node.js is required to build Angular apps
 - task: NodeTool@0
   inputs:
     versionSpec: '12.14'
   displayName: 'Install Node.js'
 
 - script: |
-    npm ci
+    ls $(Build.Repository.LocalPath)
+
+- script: |
+    npm install
+  displayName: 'npm Install'
+  workingDirectory: $(Build.Repository.LocalPath)
+
+- script: |
     npx ng build
   displayName: 'npm Build'
+  workingDirectory: $(Build.Repository.LocalPath)
 
-# skipped failing unit tests
-# https://jasmine.github.io/tutorials/your_first_suite#section-Pending_Specs
 - script: |
-    npx ng test --watch=false --browsers=ChromeHeadless
+    npx ng test --watch=false --browsers=ChromeHeadless --code-coverage
   displayName: 'Run Unit Tests'
+  workingDirectory: $(Build.Repository.LocalPath)
+
+# shopt -s globstar makes '**' search through folders of any depth
+- script: |
+    shopt -s globstar
+    npm run eslint src/**/*.ts
```

Comment:
this is relying on the shell to find the files, but eslint already has plenty of its own logic for finding files. just pass it the current directory, or `src` if you're leaving out `e2e`, and handle this via the .eslintrc and .eslintignore files. (also, not a big concern, but in some environments i don't know if i'd trust the shell to be able to handle a massive argument list if there were 1000s of files.)

Reply:
also, i think `npm run` only passes extra args to the underlying script when they're after ` -- `.

Reply:
also also - the point of the builder package, and the edits to angular.json, is so that the `ng lint` commands will just run eslint for you. so your package.json script can just be "lint": "ng lint" (or, just run npx ng lint).
and devs can also just run "ng lint" to see the issues themselves, and possibly even "ng lint --fix" to auto fix some if the underlying implementation supports it, which i think it does.

also also also - i don't think your .eslintrc follows the practices recommended by the [@angular-eslint](https://github.com/angular-eslint/angular-eslint) docs, like using `overrides`.

Reply:
I was confused earlier because I was getting a misleading error message before.
If I just tried running `npm run eslint src` it would say "No files matching the pattern "src" were found." This made me think that it either could not find the src folder even though it was in the correct directory, or, it would only work for files but not directories. Both of these initial theories were incorrect though. I eventually found this, which cleared up my understanding.

https://stackoverflow.com/a/42611291

Apparently eslint defaults to .js if you don't specify an extension, so it looked for .js files in src and couldn't find any, then said nothing matching src was found. Now that I've figured this out, I shouldn't need the shell expansion anymore.

I was able to do away with needing `npm run`, now I use `npx eslint src --ext .ts` which I think is at least a step in the right direction. But I can't figure out the error that `npx ng lint` is giving me. 

> Schema validation failed with the following errors:  Data path "" must have required property 'lintFilePatterns'.

Reply:
Got `npx ng lint` working, converted configuration to .eslintrc.json and fixed formatting.